cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -285,14 +303,18 @@ class device_row_comparator {
     __device__ element_comparator(Nullate check_nulls,
                                   column_device_view lhs,
                                   column_device_view rhs,
-                                  null_order null_precedence           = null_order::BEFORE,
-                                  int depth                            = 0,
-                                  PhysicalElementComparator comparator = {})
+                                  null_order null_precedence              = null_order::BEFORE,
+                                  int depth                               = 0,
+                                  PhysicalElementComparator comparator    = {},
+                                  dremel_device_view l_dremel_device_view = {},
+                                  dremel_device_view r_dremel_device_view = {})
```

Comment:
The `comparator` parameter is no longer the last one, as in other functions (like the one above).

Reply:
Is this a mistake?

Reply:
I think this is done intentionally because if all the pre-checks around rep/def levels succeed we then construct a new comparator that doesn't require the dremel data since it just checks value by value all the physical data at that stage. However, that still needs to forward the `PhysicalElementComparator` to the new `element_comparator`. I am still working my way through the design to understand whether there is a simpler way to approach this, but at the moment for this reason I think that it makes sense that the `element_comparator` has these parameters in a different place than the `device_row_comparator`.

However, your question does lead to a related interesting question. At the level of the `device_row_comparator` I'm not sure that the dremel parameters should really be optional. Practically speaking the preprocessed table always provides them, and our entire comparator framework essentially requires that pathway to constructing the `device_row_comparator` so I think we may just want to make them required parameters for everything except the `element_comparator`. Curious what you think.

CC @devavret in case he's available and interested.

Reply:
Following up: the `list_lex_preprocess` function creates empty device views for any columns that aren't of list type, and since the logic in the comparator just default constructs one when it encounters a non-list column I believe that it's safe to remove the optional and just always index into the `ith` element of the span. I'm removing the `optional` at the `comparator` level, but I can definitely revisit if someone finds a flaw in my reasoning.

Reply:
That aligns with my previous reading of the code. I think that’s the right path @vyasr.